workers/unity/Packages/com.improbable.gdk.core/Subscriptions/Callbacks.cs

```diff
@@ -1,9 +1,154 @@
 using System;
 using System.Collections.Generic;
-using Improbable.Gdk.Core;
 
 namespace Improbable.Gdk.Subscriptions
 {
+    internal class DescendingComparer<T> : IComparer<T> where T : IComparable<T>
+    {
+        public int Compare(T x, T y)
+        {
+            return y.CompareTo(x);
+        }
+    }
+
+    // We need to avoid changing the underlying map while iterating over it. For example, if a
+    // callback removes itself, a call to InvokeAll iterates over the map to invoke each callback.
+    // This then indirectly invokes Remove, which makes the iteration crash.
+    //
+    // A simple solution to this problem would be to copy the map for each iteration, but this
+    // allocates a significant amount of memory each time. Instead, we fix this issue by postponing
+    // changes to the map until the end of the top-level call to this object. In the mentioned
+    // example, when InvokeAll indirectly calls Remove, the removal is not done immediately, but
+    // postponed to the end of the (top-level) InvokeAll call.
+    //
+    // The current callbacks are those in the map, plus toAdd, minus toRemove. The changes are
+    // merged via the UpdateGuard.
+    internal class Callbacks<T>
```

Comment:
I'm trying to decide if it's surprising behaviour that if I indirectly `Add` a new callback in a handler, it won't be executed immediately as part of the callback cycle.
Also, if I indirectly `Remove` other callbacks in a handler, that they'll still be executed as part of the callback cycle.